File: src/net/dinkla/raytracer/materials/Reflective.java
package net.dinkla.raytracer.materials;

import net.dinkla.raytracer.brdf.GlossySpecular;
import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.hits.Shade;
import net.dinkla.raytracer.brdf.BRDF;
import net.dinkla.raytracer.brdf.PerfectSpecular;
import net.dinkla.raytracer.math.Ray;
import net.dinkla.raytracer.math.Vector3D;
import net.dinkla.raytracer.worlds.World;

public class Reflective extends Phong {

    PerfectSpecular reflectiveBrdf;

    public Reflective() {
        super();
        reflectiveBrdf = new PerfectSpecular();
    }

    public void setKr(double kr) {
        reflectiveBrdf.setKr(kr);
    }

    public void setCr(Color cr) {
        reflectiveBrdf.setCr(cr);
    }

    @Override
    public Color shade(World world, Shade sr) {
        final Color L = super.shade(world, sr);
        final Vector3D wo = sr.ray.getD().negate();
        final BRDF.Sample sample = reflectiveBrdf.sampleF(sr, wo);
        double f = sr.getNormal().dot(sample.getWi());
        final Ray reflectedRay = new Ray(sr.getHitPoint(), sample.getWi());
        final Color c1 =  world.getTracer().trace(reflectedRay, sr.depth + 1);
        final Color c2 =  sample.getColor().mult(c1).mult(f);
        return  L.plus(c2);
    }
    
}

File: src/main/kotlin/net/dinkla/raytracer/materials/Reflective.kt
package net.dinkla.raytracer.materials

import net.dinkla.raytracer.brdf.PerfectSpecular
import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.hits.Shade
import net.dinkla.raytracer.math.Ray
import net.dinkla.raytracer.utilities.hash
import net.dinkla.raytracer.world.World

class Reflective(color: Color = Color.WHITE,
                 ka: Double = 0.25,
                 kd: Double = 0.75) : Phong(color, ka, kd) {

    private val reflectiveBRDF = PerfectSpecular()

    var kr: Double
        get() = reflectiveBRDF.kr
        set(v: Double) {
            reflectiveBRDF.kr = v
        }

    var cr: Color
        get() = reflectiveBRDF.cr
        set(v: Color) {
            reflectiveBRDF.cr = v
        }

    override fun shade(world: World, sr: Shade): Color {
        val L = super.shade(world, sr)
        val wo = -sr.ray.direction
        val sample = reflectiveBRDF.sampleF(sr, wo)
        val f = sr.normal dot sample.wi
        val reflectedRay = Ray(sr.hitPoint, sample.wi)
        val c1 = world.tracer.trace(reflectedRay, sr.depth + 1)
        val c2 = sample.color * (c1).times(f)
        return L + c2
    }

    override fun equals(other: Any?): Boolean {
        if (other != null && other is Reflective) {
            return super.equals(other) && reflectiveBRDF == other.reflectiveBRDF
        }
        return false
    }

    override fun hashCode(): Int = hash(reflectiveBRDF, ambientBRDF, diffuseBRDF, specularBRDF)

    override fun toString() = "Reflective $reflectiveBRDF ${super.toString()}"
}
